File: evaluation/S1/python/Hard/1795-word-break-ii.py
# Source: https://leetcode.com/problems/word-break-ii/   |   Difficulty: Hard
#
# Problem Description:
# Given a string s and a dictionary of strings wordDict, add spaces in s to construct a sentence where each word is a valid dictionary word. Return all such possible sentences in any order.
#
# Note that the same word in the dictionary may be reused multiple times in the segmentation.
#
# Example:
# Input: s = "catsanddog", wordDict = ["cat","cats","and","sand","dog"]
# Output: ["cats and dog","cat sand dog"]
#
# Constraints:
# 1 <= s.length <= 20
# 	1 <= wordDict.length <= 1000
# 	1 <= wordDict[i].length <= 10
# 	s and wordDict[i] consist of only lowercase English letters.
# 	All the strings of wordDict are unique.
# 	Input is generated in a way that the length of the answer doesn't exceed 105.
#
# Helpful references (internal KB):
# - palindrome-partitioning (string, array, backtracking, recursion)
#   • When to use: Use this when you need to find all possible ways to break down a string into parts, where each part must satisfy a specific property. It is suitable for generating all combinations or permutations with constraints.
#   • Idea: This algorithm recursively explores all possible ways to partition a string into substrings, ensuring each substring is a palindrome. It employs a backtracking approach to build valid partitions incrementally, typically resulting in exponential time complexity.
#   • Invariants: At the start of each recursive call, the 'step' list contains a valid palindrome partition for the string segment processed so far.; The current processing index (or remaining string) always points to the unpartitioned suffix of the original input string.
#   • Tips: Utilize a helper function to manage additional state, such as the current partition path and the overall result list.; Pass the remaining string or a start index to recursive calls to define the subproblem for subsequent partitioning.
#   • Pitfalls: Forgetting to create a new copy of the current partition path before adding it to the final result list, leading to incorrect results.; Incorrectly defining the base case for recursion, which can cause infinite loops or missed valid partitions.
# - Introduction to Dynamic Programming (array, tree, recursion, dp-1d)
#   • When to use: Use when a problem has overlapping subproblems and optimal substructure, and a natural recursive solution is inefficient due to redundant computations. It's ideal for transforming exponential recursive solutions into polynomial time.
#   • Idea: Top-down dynamic programming, or memoization, optimizes recursive solutions by storing the results of expensive function calls and returning the cached result when the same inputs occur again. This technique typically reduces time complexity from exponential to polynomial, often O(N) or O(N*M).
#   • Invariants: memo[state] contains the computed result for 'state' if memo[state] is not the sentinel value.; If memo[state] is the sentinel value, the result for 'state' has not yet been computed.
#   • Tips: Initialize the memoization table with a sentinel value (e.g., -1, null) to distinguish uncomputed states from valid zero/false results.; Use arrays for lookup tables when the state space is contiguous and small for optimal cache performance.
#   • Pitfalls: Forgetting to initialize the memoization table, leading to incorrect results or infinite loops.; Incorrectly identifying or handling base cases in the recursive function.
# - Aho-Corasick Algorithm for Pattern Searching (string, array, aho-corasick, enumeration)
#   • When to use: Use when you need to find all occurrences of a fixed set of patterns (a dictionary) within a larger text efficiently. It is particularly effective for scenarios with many patterns or long texts where patterns might overlap.
#   • Idea: The Aho-Corasick algorithm constructs a finite automaton from a dictionary of patterns, then uses this automaton to search for all pattern occurrences in a text in a single pass. It achieves a time complexity of O(N + M + K) where N is text length, M is total pattern length, and K is total matches.
#   • Invariants: The trie structure correctly represents all prefixes of all patterns in the dictionary.; Each node's failure link points to the longest proper suffix of the string ending at that node which is also a prefix in the trie.
#   • Tips: Build the trie first from all patterns, marking terminal nodes.; Compute failure links using a BFS-like approach, similar to KMP's LPS array.
#   • Pitfalls: Incorrectly building failure links can lead to missed matches or infinite loops.; Forgetting to handle output links (dictionary links) will miss patterns found via failure transitions.
#
# Notes:
# - Keep the given function signature unchanged.
# - Use only the Python standard library; no third-party packages.
# - Do not change input/output formats or return value semantics as defined by the template/problem.
# - Conformance: use features compatible with LeetCode’s Python3 runtime; do NOT rely on undefined/implementation-specific behavior.
# - If you introduce any helper functions/classes, include their full implementations in this file; no hidden helpers.

class Solution(object):
    def wordBreak(self, s, wordDict):
        """
        :type s: str
        :type wordDict: List[str]
        :rtype: List[str]
        """
        # Step 1: Create a set of words for O(1) lookups
        word_set = set(wordDict)
        
        # Step 2: Use a memoization dictionary to store results for subproblems
        memo = {}
        
        def backtrack(start):
            # If we have already computed the result for this start index, return it
            if start in memo:
                return memo[start]
            
            # If we have reached the end of the string, return a list with an empty string
            if start == len(s):
                return [""]

            # Initialize a list to store sentences formed from this start index
            sentences = []
            
            # Explore all possible end indices for the current start index
            for end in range(start + 1, len(s) + 1):
                word = s[start:end]
                # If the current substring is in the word set, recursively backtrack from the end index
                if word in word_set:
                    for sub_sentence in backtrack(end):
                        if sub_sentence:
                            sentences.append(word + " " + sub_sentence)
                        else:
                            sentences.append(word)

            # Store the computed sentences in the memoization dictionary before returning
            memo[start] = sentences
            return sentences
        
        # Start backtracking from index 0
        return backtrack(0)

File: evaluation/S0/c/Hard/1795-word-break-ii.c
// Source: https://leetcode.com/problems/word-break-ii/   |   Difficulty: Hard
//
// Problem Description:
// Given a string s and a dictionary of strings wordDict, add spaces in s to construct a sentence where each word is a valid dictionary word. Return all such possible sentences in any order.
//
// Note that the same word in the dictionary may be reused multiple times in the segmentation.
//
// Example:
// Input: s = "catsanddog", wordDict = ["cat","cats","and","sand","dog"]
// Output: ["cats and dog","cat sand dog"]
//
// Constraints:
// 1 <= s.length <= 20
// 	1 <= wordDict.length <= 1000
// 	1 <= wordDict[i].length <= 10
// 	s and wordDict[i] consist of only lowercase English letters.
// 	All the strings of wordDict are unique.
// 	Input is generated in a way that the length of the answer doesn't exceed 105.
//

/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
char** wordBreak(char* s, char** wordDict, int wordDictSize, int* returnSize) {

    *returnSize = 0;
    int sLen = strlen(s);
    
    // DP to check if s can be segmented
    int* dp = (int*)calloc(sLen + 1, sizeof(int));
    dp[0] = 1;
    
    for (int i = 1; i <= sLen; i++) {
        for (int j = 0; j < wordDictSize; j++) {
            int wLen = strlen(wordDict[j]);
            if (i >= wLen && dp[i - wLen] && strncmp(&s[i - wLen], wordDict[j], wLen) == 0) {
                dp[i] = 1;
                break;
            }
        }
    }
    
    if (!dp[sLen]) {
        free(dp);
        return NULL;
    }
    
    free(dp);
    
    // Backtracking to find all sentences
    char** results = NULL;
    int resultsCap = 0;
    
    void backtrack(int start, char* path, int pathLen) {
        if (start == sLen) {
            if (*returnSize >= resultsCap) {
                resultsCap = resultsCap == 0 ? 16 : resultsCap * 2;
                results = (char**)realloc(results, resultsCap * sizeof(char*));
            }
            path[pathLen - 1] = '\0'; // Replace last space with null terminator
            results[*returnSize] = strdup(path);
            (*returnSize)++;
            return;
        }
        
        for (int i = start + 1; i <= sLen; i++) {
            for (int j = 0; j < wordDictSize; j++) {
                int wLen = strlen(wordDict[j]);
                if (i - start == wLen && strncmp(&s[start], wordDict[j], wLen) == 0) {
                    memcpy(&path[pathLen], wordDict[j], wLen);
                    path[pathLen + wLen] = ' ';
                    backtrack(i, path, pathLen + wLen + 1);
                }
            }
        }
    }
    
    char* path = (char*)malloc((sLen + wordDictSize) * sizeof(char)); // Extra space for spaces
    backtrack(0, path, 0);
    free(path);
    
    return results;
            
}
